File: packages/whisker-input/ios/Sources/WhiskerInput/InputModule.swift
// `whisker-input` ModuleDefinition (iOS).
//
// The SwiftPM build plugin discovers this Module subclass and emits its
// Whisker registration entry. Android uses the same ModuleDefinition shape.

import WhiskerModule

@WhiskerModule
public final class InputModule: Module {
    public override func definition() -> ModuleDefinition {
        ModuleDefinition {
            Name("Input")
            View("whisker-input:Input", WhiskerInputView.self) {

                // ---- Value + placeholder ----------------------------------

                Prop("value") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setValue(value.asString ?? "")
                }
                Prop("placeholder") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setPlaceholder(value.asString ?? "")
                }
                Prop("placeholder-color") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setPlaceholderColor(value)
                }

                // ---- Cursor / selection colours --------------------------

                Prop("caret-color") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setCaretColor(value)
                }
                Prop("selection-color") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setSelectionColor(value)
                }

                // ---- Layout mode -----------------------------------------

                Prop("multiline") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setMultiline(value.asString ?? "false")
                }
                Prop("lines") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setLines(value.asString ?? "0")
                }

                // ---- Input behaviour -------------------------------------

                Prop("secure") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setSecure(value.asString ?? "false")
                }
                Prop("editable") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setEditable(value.asString ?? "true")
                }
                Prop("auto-focus") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setAutoFocus(value.asString ?? "false")
                }
                Prop("max-length") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setMaxLength(value.asString ?? "0")
                }

                // ---- Keyboard / return key -------------------------------

                Prop("keyboard-type") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setKeyboardType(value.asString ?? "default")
                }
                Prop("return-key") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setReturnKey(value.asString ?? "default")
                }
                Prop("auto-capitalize") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setAutoCapitalize(value.asString ?? "sentences")
                }
                Prop("autocorrect") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setAutocorrect(value.asString ?? "true")
                }
                Prop("spell-check") { (view: WhiskerInputView, value: WhiskerValue) in
                    view.setSpellCheck(value.asString ?? "true")
                }

                TextStyle { (view: WhiskerInputView, style: WhiskerTextStyle) in
                    view.applyTextStyle(style)
                }

                // Declaration-only metadata for the codegen / docs scanner;
                // dispatch happens in `InputView.swift`.
                Events("input", "change", "focus", "blur", "submit")

                // ---- One-way View commands -------------------------------

                Command("focus") { (view: WhiskerInputView, _: WhiskerValue) in
                    view.focusField()
                }
                Command("blur") { (view: WhiskerInputView, _: WhiskerValue) in
                    view.blurField()
                }
                Command("clear") { (view: WhiskerInputView, _: WhiskerValue) in
                    view.clearField()
                }
                Command("setValue") { (view: WhiskerInputView, parameters: WhiskerValue) in
                    if case .map(let m) = parameters, let s = m["value"]?.asString {
                        view.setValue(s)
                    }
                }
            }
        }
    }
}

File: packages/whisker-input/android/src/main/kotlin/rs/whisker/elements/input/InputModule.kt
// `whisker-input` ModuleDefinition (Android). KSP discovers the annotated
// module and generates its registration entry.

package rs.whisker.elements.input

import rs.whisker.runtime.Module
import rs.whisker.runtime.WhiskerModule
import rs.whisker.runtime.ModuleDefinition
import rs.whisker.runtime.WhiskerValue


@WhiskerModule
class InputModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("Input")
        View("whisker-input:Input", WhiskerInputView::class.java) {
            // ---- text-content props ----------------------------------------

            Prop("value") { view: WhiskerInputView, value ->
                view.setValue(value.asString() ?: "")
            }
            Prop("placeholder") { view: WhiskerInputView, value ->
                view.setPlaceholder(value.asString() ?: "")
            }
            Prop("placeholder-color") { view: WhiskerInputView, value ->
                view.setPlaceholderColor(value.asString() ?: "")
            }
            Prop("caret-color") { view: WhiskerInputView, value ->
                view.applyCaretColor(value.asString() ?: "")
            }
            Prop("selection-color") { view: WhiskerInputView, value ->
                view.setSelectionColor(value.asString() ?: "")
            }

            // ---- behaviour props -------------------------------------------

            Prop("multiline") { view: WhiskerInputView, value ->
                view.setMultiline(value.asString() ?: "false")
            }
            Prop("lines") { view: WhiskerInputView, value ->
                view.setLines(value.asString() ?: "0")
            }
            Prop("secure") { view: WhiskerInputView, value ->
                view.setSecure(value.asString() ?: "false")
            }
            Prop("editable") { view: WhiskerInputView, value ->
                view.setEditable(value.asString() ?: "true")
            }
            Prop("auto-focus") { view: WhiskerInputView, value ->
                view.setAutoFocus(value.asString() ?: "false")
            }
            Prop("max-length") { view: WhiskerInputView, value ->
                view.setMaxLength(value.asString() ?: "0")
            }
            Prop("keyboard-type") { view: WhiskerInputView, value ->
                view.setKeyboardType(value.asString() ?: "default")
            }
            Prop("return-key") { view: WhiskerInputView, value ->
                view.setReturnKey(value.asString() ?: "default")
            }
            Prop("auto-capitalize") { view: WhiskerInputView, value ->
                view.setAutoCapitalize(value.asString() ?: "sentences")
            }
            Prop("autocorrect") { view: WhiskerInputView, value ->
                view.setAutocorrect(value.asString() ?: "true")
            }
            Prop("spell-check") { view: WhiskerInputView, value ->
                view.setSpellCheck(value.asString() ?: "true")
            }

            // Declaration-only metadata (parity with the iOS module);
            // actual dispatch is the imperative WhiskerCustomEvent path
            // inside WhiskerInputView. Documents the emittable set.
            Events("input", "change", "focus", "blur", "submit")

            TextStyle { view: WhiskerInputView, style ->
                view.applyTextStyle(style)
            }

            // ---- one-way View commands -------------------------------------

            Command("focus") { view: WhiskerInputView, _ ->
                view.focusField()
            }
            Command("blur") { view: WhiskerInputView, _ ->
                view.blurField()
            }
            // `clear` fires `input` so the bound signal sees the change as
            // though the user had typed it.
            Command("clear") { view: WhiskerInputView, _ ->
                view.clearField()
            }
            // The view applies the cursor-diff guard and suppresses the
            // resulting afterTextChanged, which is not a user edit.
            Command("setValue") { view: WhiskerInputView, parameters ->
                val map = (parameters as? WhiskerValue.Map)?.value
                val text = map?.get("value")?.asString() ?: ""
                view.setValueExternal(text)
            }
        }
    }
}
